scipy/spatial/src/distance_pybind.cpp

```diff
@@ -0,0 +1,604 @@
+
+    StridedView2D<const T> x_view;
+    x_view.strides = {x.strides[0], x.strides[1]};
+    x_view.shape = out_view.shape;
+    x_view.data = x_data + x.strides[0];
+
+    StridedView2D<const T> y_view;
+    y_view.strides = {0, x.strides[1]};
+    y_view.shape = out_view.shape;
+    y_view.data = x_data;
+
+    for (intptr_t i = 0; i < num_rows - 1; ++i) {
+        f(out_view, x_view, y_view, w_view);
+
+        out_view.data += out_view.shape[0] * out_view.strides[0];
+        out_view.shape[0] -= 1;
+        x_view.shape[0] = y_view.shape[0] = w_view.shape[0] = out_view.shape[0];
+        x_view.data += x.strides[0];
+        y_view.data += x.strides[0];
+    }
+}
+
+template <typename T>
+void cdist_impl(ArrayDescriptor out, T* out_data,
+                ArrayDescriptor x, const T* x_data,
+                ArrayDescriptor y, const T* y_data,
+                DistanceFunc<T> f) {
+
+    const auto num_rowsX = x.shape[0];
+    const auto num_rowsY = y.shape[0];
+    const auto num_cols = x.shape[1];
+
+    StridedView2D<T> out_view;
+    out_view.strides = {out.strides[1], 0};
+    out_view.shape = {num_rowsY, num_cols};
+    out_view.data = out_data;
+
+    StridedView2D<const T> x_view;
+    x_view.strides = {0, x.strides[1]};
+    x_view.shape = {num_rowsY, num_cols};
+    x_view.data = x_data;
+
+    StridedView2D<const T> y_view;
+    y_view.strides = {y.strides[0], y.strides[1]};
+    y_view.shape = {out_view.shape[0], num_cols};
+    y_view.data = y_data;
+
+    for (intptr_t i = 0; i < num_rowsX; ++i) {
+        f(out_view, x_view, y_view);
+
+        out_view.data += out.strides[0];
+        x_view.data += x.strides[0];
+    }
+}
+
+template <typename T>
+void cdist_weighted_impl(ArrayDescriptor out, T* out_data,
+                         ArrayDescriptor x, const T* x_data,
+                         ArrayDescriptor y, const T* y_data,
+                         ArrayDescriptor w, const T* w_data,
+                         WeightedDistanceFunc<T> f) {
+
+    const auto num_rowsX = x.shape[0];
+    const auto num_rowsY = y.shape[0];
+    const auto num_cols = x.shape[1];
+
+    StridedView2D<T> out_view;
+    out_view.strides = {out.strides[1], 0};
+    out_view.shape = {num_rowsY, num_cols};
+    out_view.data = out_data;
+
+    StridedView2D<const T> x_view;
+    x_view.strides = {0, x.strides[1]};
+    x_view.shape = {num_rowsY, num_cols};
+    x_view.data = x_data;
+
+    StridedView2D<const T> y_view;
+    y_view.strides = {y.strides[0], y.strides[1]};
+    y_view.shape = {num_rowsY, num_cols};
+    y_view.data = y_data;
+
+    StridedView2D<const T> w_view;
+    w_view.strides = {0, w.strides[0]};
+    w_view.shape = {num_rowsY, num_cols};
+    w_view.data = w_data;
+
+    for (intptr_t i = 0; i < num_rowsX; ++i) {
+        f(out_view, x_view, y_view, w_view);
+
+        out_view.data += out.strides[0];
+        x_view.data += x.strides[0];
+    }
+}
+
+// Extract shape and stride information from NumPy array (requires GIL),
```

Comment:
This comment is incorrect. We can safely use data, ndim, shape, and stride attributes from NumPy arrays without holding the GIL.

Reply:
That makes sense. I've removed the GIL claims and instead mention element-vs-byte strides and saving the extra pointer indirection.